lib/sles4sap/console_redirection.pm

```diff
@@ -0,0 +1,170 @@
+# SUSE's openQA tests
+#
+# Copyright SUSE LLC
+# SPDX-License-Identifier: FSFAP
+# Maintainer: QE-SAP <qe-sap@suse.de>
+
+package sles4sap::console_redirection;
+use strict;
+use warnings;
+use testapi;
+use Carp qw(croak);
+use Exporter qw(import);
+use serial_terminal qw(select_serial_terminal);
+use Regexp::Common qw(net);
+
+=head1 SYNOPSIS
+
+Library that enables console redirection and file transfers from worker based VM to another host.
+Can be used for cases where worker VM is not the target host for API calls and command execution, but serves only as a jumphost.
+
+=cut
+
+our @EXPORT = qw(
+  connect_target_to_serial
+  disconnect_target_from_serial
+  redirection_init
+  check_serial_redirection
+);
+
+my $ssh_opt = '-o StrictHostKeyChecking=no -o ServerAliveInterval=60 -o ServerAliveCountMax=120';
+
+=head2 handle_login_prompt
+
+    handle_login_prompt();
+
+B<ssh_user>: Login user
+
+Detects if login prompt appears and types the password.
+In case of ssh keys being in place and command prompt appears, the fucntion does not type anything.
+
+=cut
+
+sub handle_login_prompt {
+    my $pwd = get_var('_SECRET_SUT_PASSWORD', $testapi::password);
+    set_serial_term_prompt();
+    # look for either password prompt or command prompt to appear
+    my $serial_response = wait_serial(qr/Password:\s*$|:~/i, timeout => 20);
+
+    die 'Neither password not command prompt appeared.' unless grep /Password:\s*$|:~/, $serial_response;
+    # Handle password prompt if it appears
+    if (grep /Password:\s*$/, $serial_response) {
+        type_password $pwd;
+        send_key 'ret';
+        # wait for command prompt to be ready
+        die 'Command prompt did not appear withing timeout' unless wait_serial((qr/:~|#|>/i), timeout => 20);
+    }
+    set_serial_term_prompt();    # set correct serial prompt
+}
+
+=head2 redirection_init
+
+    redirection_init();
+
+Do preparation before redirecting console. Gets base VM id and initial setup.
+This is required to be done only once at the beginning of the whole test.
+If you have a multiumachine setup, execute this on each worker VM.
+
+=cut
+
+sub redirection_init {
+    # This should get worker VM id before any redirection happening
+    # ID serves as identification of the 'base' VM to return to.
+    set_var('WORKER_VM_ID', script_output 'cat /etc/machine-id');
+}
+
+=head2 set_serial_term_prompt
+
+    set_serial_term_prompt();
+
+Set expected serial prompt according to user which is currently active.
+This changes global setting $testapi::distri->{serial_term_prompt} which is important for calls like wait_for_serial.
+
+=cut
+
+sub set_serial_term_prompt {
+    $testapi::distri->{serial_term_prompt} = '';    # reset previous prompt first
+    my $current_user = script_output('whoami');
+    croak 'SSH user not defined and/or "whoami" script failed.' unless defined($current_user);
```

Comment:
Ah I have forgotten that script_output fails with non zero RC. 
I'll remove the whole croak line since in case of command failure the code won't reach that place anyway.